cloud-sql/mysql/Program.cs

```diff
@@ -14,21 +14,68 @@
  * the License.
  */
 
+using CloudSql.Settings;
+using Google.Cloud.Diagnostics.AspNetCore;
 using Microsoft.AspNetCore;
 using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Hosting;
+using System;
+using System.IO;
 
 namespace CloudSql
 {
     public class Program
     {
+        public static AppSettings AppSettings { get; private set; }
+
         public static void Main(string[] args)
         {
-            BuildWebHost(args).Run();
+            BuildWebHost(args).Build().Run();
         }
 
-        public static IWebHost BuildWebHost(string[] args) =>
-            WebHost.CreateDefaultBuilder(args)
+        public static IWebHostBuilder BuildWebHost(string[] args)
+        {
+            ReadAppSettings();
+
+            return WebHost.CreateDefaultBuilder(args)
+                .UseGoogleDiagnostics(
+                    AppSettings.GoogleCloudSettings.ProjectId,
+                    AppSettings.GoogleCloudSettings.ServiceName,
+                    AppSettings.GoogleCloudSettings.Version)
                 .UseStartup<Startup>()
+                .UsePortEnvironmentVariable();
+        }
+
+        /// <summary>
+        /// Read application settings from appsettings.json. 
+        /// </summary>
+        private static void ReadAppSettings()
+        {
+            var config = new ConfigurationBuilder()
+                .SetBasePath(Directory.GetCurrentDirectory())
+                .AddJsonFile("appsettings.json", optional: false)
                 .Build();
+
+            // Read json config into AppSettings.
+            AppSettings = new AppSettings();
+            config.Bind(AppSettings);
+        }
+    }
+
+    static class ProgramExtensions
+    {
+        // Google Cloud Run sets the PORT environment variable to tell this
+        // process which port to listen to.
+        public static IWebHostBuilder UsePortEnvironmentVariable(
+            this IWebHostBuilder builder)
+        {
+            string port = Environment.GetEnvironmentVariable("PORT");
+            if (!string.IsNullOrEmpty(port))

```

Comment:
Should you add and `else` here that uses `8080` if the port is not set on an environment variable?

Reply:
I just tested updating this line to be:
string port = Environment.GetEnvironmentVariable("PORT") ?? "8080";

When I start the cloudsql proxy locally and run the program after that change it starts up successfully with the command line message:

"Application started. Press Ctrl+C to shut down.
Now listening on: http://0.0.0.0:8080"

But when I visit "http://0.0.0.0:8080" in a browser I get the error message:

"This site can’t be reachedThe webpage at http://0.0.0.0:8080/ might be temporarily down or it may have moved permanently to a new web address. ERR_ADDRESS_INVALID"


Changing the line back to: 
string port = Environment.GetEnvironmentVariable("PORT");

When I start the cloudsql proxy locally and run the program after that change it
starts up successfully with the command line message:

"Now listening on: http://localhost:5000
Now listening on: https://localhost:5001
Application started. Press Ctrl+C to shut down."

Now when visit "http://localhost:5000" or "https://localhost:5001/" in a browser I see the web app load correctly in a functional state.

I was also able to successfully deploy and run the app (with the current port setting command) to Cloud Run and App Engine.

This is also currently being used for the getting-started sample app:
https://github.com/GoogleCloudPlatform/getting-started-dotnet/blob/master/HelloWorld/Program.cs

Please force merge if everything else looks okay. Thanks!

Reply:
OK! That sounds like something else is setting a default port then which is fine.
I'll let tests run now (I've deactivated IAP) and then you can merge on green.